cpp/ql/src/semmlecode.cpp.dbscheme

```diff
@@ -755,6 +760,11 @@ function_template_argument(
     int index: int ref,
     int arg_type: @type ref
 );
+function_template_argument_value(
+    int function_id: @function ref,
+    int index: int ref,
+    int arg_value: @expr ref
```

Comment:
I'd like to see some practical examples of what this `@expr` refers to. I'd like to see how this gets represented:
```cpp
template<int x>
struct C { };

static const int one1 = 1, one2 = 1;

C<one1> c = C<one2>();
```

To fix issues like [this user-reported FP](https://discuss.lgtm.com/t/false-positive-about-unused-static-constexpr-integer/2238) we need `VariableAccess`es to both `one1` and `one2`, but I'd also expect the actual type of the instantiated class to be `C<1>` in both cases so we don't get two copies. This is analogous to resolving typedefs when instantiating templates with type parameters, which I think we do today.

Reply:
This still doesn't reference the variables.  I will investigate.

Reply:
I don't think we can do this with the current extractor infrastruture.  When we get to see the template parameters the front-end has expanded `one1` and `one2` into `1` and thrown away the source info.  This is understandable `C<one1>` and `C<one2>` are the same instantiation of the template even though they look different in the source.

Reply:
It sounds correct to me that the instantiated template is `C<1>` since that's how most types work in our db: a `Type` does not have a location because it exists in only one copy even if it's referred to from many locations. Is there only a single copy of the `1` expression even if the `C<1>` is mentioned multiple times? Does the `1` expression have a location or a parent? What if it's not an integer `1` but something more complex like `&myGlobalVariable`?

To fix the FP, we still need a `VariableAccess` to `one1` and `one2`. Even if it's just a free-floating expression with no parent, that should be good enough for the "unused static variable" query.

To be clear, I don't expect this PR to solve all problems we have with non-type template parameters. I'm happy to see it merged as it is (with a few more tests) as long as the QLDoc explains clearly how things work.

Reply:
See: https://github.com/Semmle/ql/pull/2153/files#diff-762d2efe4797240694b1336bc38d834b.

But the summary is: there is no way for us to know that `one1` and `one2` are involved in instantiating the template becuase the FE just gives us the constant result.